src/protocol/prss.rs

```diff
@@ -153,59 +150,91 @@ impl Participant {
     /// # Panics
     /// When used incorrectly.  For instance, if you ask for an RNG and then ask
     /// for a PRSS using the same key.
-    pub fn prss(&self, key: impl AsRef<str>) -> &PrssSpace {
-        let p = self.inner.lock().unwrap().prss(key.as_ref());
-        // Safety: This pointer refers to pinned memory that is allocated
-        // and retained as long as this object exists.
-        unsafe { p.as_ref() }
+    pub fn indexed(&self, key: &UniqueStepId) -> Arc<IndexedSharedRandomness> {
+        self.inner.lock().unwrap().indexed(key.as_ref())
+    }
+
+    /// Get a sequential shared randomness.
+    ///
+    /// # Panics
+    /// This can only be called once.  After that, calls to this function or `indexed` will panic.
+    pub fn sequential(
+        &self,
+        key: &UniqueStepId,
+    ) -> (SequentialSharedRandomness, SequentialSharedRandomness) {
+        self.inner.lock().unwrap().sequential(key.as_ref())
     }
 }
 
-impl Debug for Participant {
+impl Debug for Endpoint {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.write_str("Participant")
     }
 }
 
-struct ParticipantInner {
+enum EndpointItem {
+    Indexed(Arc<IndexedSharedRandomness>),
+    Sequential,
+}
+
+struct EndpointInner {
     left: GeneratorFactory,
     right: GeneratorFactory,
     // TODO(mt): add a function to get an RNG instead of the indexed PRSS.
     // That should mark the entry as dead, so that any attempt to get the
     // indexed PRSS or another RNG will fail.
-    spaces: HashMap<String, Pin<Box<PrssSpace>>>,
+    items: HashMap<String, EndpointItem>,
 }
 
-impl ParticipantInner {
-    pub fn prss(&mut self, key: &str) -> NonNull<PrssSpace> {
+impl EndpointInner {
+    pub fn indexed(&mut self, key: &str) -> Arc<IndexedSharedRandomness> {
         // The second arm of this statement would be fine, except that `HashMap::entry()`
         // only takes an owned value as an argument.
         // This makes the lookup perform an allocation, which is very much suboptimal.
-        let p = if let Some(value) = self.spaces.get(key) {
-            value
+        let item = if let Some(item) = self.items.get(key) {
+            item
         } else {
-            self.spaces.entry(key.to_owned()).or_insert_with_key(|k| {
-                Box::pin(PrssSpace {
+            self.items.entry(key.to_owned()).or_insert_with_key(|k| {
+                EndpointItem::Indexed(Arc::new(IndexedSharedRandomness {
```

Comment:
i still feel a little nervous about keeping all prss instances forever, but I guess we can change it later?

Reply:
They will be reclaimed once the query is over, so it's not so bad.  We don't otherwise have any signal we can use to remove them, but if we ever get one, then we can do something about it.